pkg/controller/azure/provider/resourcegroup.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2018 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/crossplaneio/crossplane/pkg/apis/azure/v1alpha1"
+	azureclient "github.com/crossplaneio/crossplane/pkg/clients/azure"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerNameRG           = "azure.resourcegroup"
+	errorCreatingClientRG      = "Failed to create Azure client"
+	errorTestingClientRG       = "Failed testing Azure client"
+	errorFetchingAzureProvider = "failed to fetch Azure Provider"
+)
+
+// ResourceGroupReconciler reconciles a Provider object
+type ResourceGroupReconciler struct {
+	client.Client
+	scheme     *runtime.Scheme
+	kubeclient kubernetes.Interface
+	recorder   record.EventRecorder
+
+	create func(*azureclient.Client, string, string) error
+}
```

Comment:
Do we also need `sync` and `delete` functionality on this controller? There's not a whole lot to sync on with a resource group, but I do notice they have a `provisioningState` which I assume reflects whether they're ready for use - we may want to sync that.

Reply:
So I had some questions about a `delete` function because deleting an Azure Resource Group requires that you move or delete all resources within it. Do we want the functionality of deleting a Resource Group in Crossplane to only allow deletion if it is empty?

Reply:
I would vote for having the full functionality in this resource group controller that the other controllers have, so including a sync and delete function makes sense to me.  This may feel a bit odd since a resource group has somewhat of a different life-cycle compared to other objects like databases and buckets, but I think it's the right place to start for completeness.  We should be able to clean up all objects that we create, and we should be able to sync changes in between the cloud provider and the k8s API object.

I assume that the Azure API will return a 403 - BadRequest or something like that if you try to delete a resource group that has resources still alive in it.  It would be good to understand what that behavior is.  If that is the case, then this looks similar to a typical controller reconciliation pattern where the controller keeps trying to drive the actual state towards the user's desired state.  In this case, that means it would keep trying to delete the resource group until it succeeds.

Is this the most efficient way to handle this?  Probably not, but I think that in this PR that doing `create`, `sync`, and `delete` is a great place to start.  We can refine this later on as we get more hands on experience with managing the life-cycles of resource groups.

Reply:
> Do we want the functionality of deleting a Resource Group in Crossplane to only allow deletion if it is empty?

There are two existing patterns in Kubernetes that may inform this. When you delete a `Namespace` all of its contents are automatically deleted. There may be a knob to change this, but it is in my experience the default behaviour. A safer alternative is the pattern used when a `PersistentVolumeClaim` is deleted while still in use by a `Pod`. The `PersistentVolumeClaim` changes state to `Terminating`, but continues to function as normal until the consuming `Pod` terminates. This would be closer to what @jbw976 suggested above.